src/index.ts

```diff
@@ -1,70 +1,109 @@
+    if (!options.canvas) {
+      throw new Error('canvas element is required!');
+    }
+
+    // global context of all modules.
+    this.context = this.initalizeContext(options);
+
+    // dispatcher role in flux architecture.
+    this.renderer = new Renderer(
+      options.painter || new Painter(this.context),
+      new Rasterizer(this.context),
+    );
+
+    // trigger events (ex. mouse event)
+    this.trigger = new Trigger(this.context);
   }
 
-  private createViewport(options?: IntializeOptions): Viewport {
-    const id: string = options.id || 'boardy';
-    const containerElement: Element = options.containerElement ||
-      window.document.body;
-    const size: [number, number] = options.size || [500, 500];
-    const resolution: [number, number] = options.resolutionWidth ?
-      [options.resolutionWidth, (options.resolutionWidth / size[0]) * size[1]] :
-      [100000, (100000 / size[0]) * size[1]];
-    const unit: number = size[0] > size[1] ?
-      resolution[1] / size[1] :
-      resolution[0] / size[0];
+  public on(handler: (action: Action) => void) {
+    this.trigger.on(handler);
+  }
+
+  public render(action: Action) {
+    this.renderer.render(action);
+  }
+
+  /**
+   * options => context
+   *   + offscreen canvas
+   *   + unit
+   */
+  private initalizeContext(options?: IntializeOptions): Context {
+    const screenElement: HTMLCanvasElement = options.canvas;
+    const offscreenElement: HTMLCanvasElement = document.createElement('canvas');
+    const size: [number, number] = [
+      screenElement.offsetWidth,
+      screenElement.offsetHeight,
+    ] || [500, 500];
+    const resolution: [number, number] = this.initializeResolution(options, size);
+    const unit: number = this.calculateUnit(size, resolution);
+
+    offscreenElement.width = resolution[0];
+    offscreenElement.height = resolution[1];
 
     return {
-      id,
-      containerElement,
-      svgElement: this.createSvgElement(id, containerElement, size, resolution),
+      canvas: {
+        screen: screenElement,
+        offscreen: offscreenElement,
+      },
+      ctx: {
+        screen: screenElement.getContext('2d'),
+        offscreen: offscreenElement.getContext('2d'),
+      },
       size,
       resolution,
       unit,
     };
   }
 
-  private createSvgElement(
-    id: string,
-    containerElement: Element,
+  private initializeResolution(
+    options: IntializeOptions,
     size: [number, number],
-    resolution: [number, number],
-  ) {
-    if (!containerElement) {
-      throw new Error('container element is required !');
+  ): [number, number] {
+    const resolution: [number, number] = [0, 0];
+
+    if (options.resolution) {
+      resolution[0] = options.resolution.width ||
+        (options.resolution.height / size[1]) * size[0];
+      resolution[1] = options.resolution.height ||
+        (options.resolution.width / size[0]) * size[1];
+    } else {
+      resolution[0] = 2048;
```

Comment:
why did you set default as 2048?

Reply:
2048 means resolution pixel count.
2048 pixels can cover most modern devices resolution.